crates/bevy_app/src/plugin_group.rs

```diff
@@ -357,23 +379,105 @@ impl PluginGroupBuilder {
-    /// be a plugin of type `Target` in the group or it will panic.
-    pub fn add_before<Target: Plugin>(mut self, plugin: impl Plugin) -> Self {
-        let target_index = self.index_of::<Target>();
-        self.order.insert(target_index, type_id_of_val(&plugin));
+    ///
+    /// If the plugin was already the group, it is removed from its previous place.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `Target` is not already in this [`PluginGroupBuilder`].
+    pub fn add_before<Target: Plugin>(self, plugin: impl Plugin) -> Self {
+        self.try_add_before_overwrite::<Target, _>(plugin)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Plugin does not exist in group: {}.",
+                    core::any::type_name::<Target>()
+                )
+            })
+    }
+
+    /// Adds a [`Plugin`] in this [`PluginGroupBuilder`] before the plugin of type `Target`.
+    ///
+    /// If the plugin was already in the group the add fails. If there isn't a plugin
+    /// of type `Target` in the group the plugin we're trying to insert is returned.
+    pub fn try_add_before<Target: Plugin, Insert: Plugin>(
+        self,
+        plugin: Insert,
+    ) -> Result<Self, (Self, Insert)> {
+        if self.contains::<Insert>() {
+            return Err((self, plugin));
+        }
+
+        self.try_add_before_overwrite::<Target, _>(plugin)
+    }
+
+    /// Adds a [`Plugin`] in this [`PluginGroupBuilder`] before the plugin of type `Target`.
+    ///
+    /// If the plugin was already in the group, it is removed from its previous places.
+    /// If there isn't a plugin of type `Target` in the group the plugin we're trying to insert
+    /// is returned.
+    pub fn try_add_before_overwrite<Target: Plugin, Insert: Plugin>(
+        mut self,
+        plugin: Insert,
+    ) -> Result<Self, (Self, Insert)> {
+        let Some(target_index) = self.index_of::<Target>() else {
+            return Err((self, plugin));
+        };
+
+        self.order.insert(target_index, TypeId::of::<Insert>());
         self.upsert_plugin_state(plugin, target_index);
-        self
+        Ok(self)
     }
 
     /// Adds a [`Plugin`] in this [`PluginGroupBuilder`] after the plugin of type `Target`.
-    /// If the plugin was already the group, it is removed from its previous place. There must
-    /// be a plugin of type `Target` in the group or it will panic.
-    pub fn add_after<Target: Plugin>(mut self, plugin: impl Plugin) -> Self {
-        let target_index = self.index_of::<Target>() + 1;
-        self.order.insert(target_index, type_id_of_val(&plugin));
+    ///
+    /// If the plugin was already the group, it is removed from its previous place.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `Target` is not already in this [`PluginGroupBuilder`].
+    pub fn add_after<Target: Plugin>(self, plugin: impl Plugin) -> Self {
+        self.try_add_after_overwrite::<Target, _>(plugin)
+            .unwrap_or_else(|_| {
```

Comment:
Isn't this just `expect()`?

Reply:
Not quite! The unwrap_or_else pattern is useful as a performance optimization when string formatting needs to be done in the error case. IIRC otherwise things are pessimized even in the happy case.